Added Points to lines c++ processing algorithm

uclaros opened the pull request:
## Description
This is a c++ processing algorithm that creates line layers from point layers, similar to the existing python *Points to path* but with extended functionality:
- Added support for *natural sorting* ( '10' > '9' )
- Avoid creating invalid lines with a single vertex
- Avoid creating closed lines with two vertices

Tests will follow

Supersedes #38701
Addresses #39596, #35525

<!--
  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

uclaros committed:
Added Points to lines algorithm

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstolines.cpp:
```diff
@@ -0,0 +1,226 @@
+
+QString QgsPointsToLinesAlgorithm::name() const
+{
+  return QStringLiteral( "pointstolines" );
```
Can we keep the name as pointstopath, and remove the existing python algorithm please?

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstolines.cpp:
```diff
@@ -0,0 +1,226 @@
+
+QString QgsPointsToLinesAlgorithm::group() const
+{
+  return QObject::tr( "Vector geometry" );
```
```suggestion
  return QObject::tr( "Vector creation" );
```

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstolines.cpp:
```diff
@@ -0,0 +1,226 @@
+
+QString QgsPointsToLinesAlgorithm::groupId() const
+{
+  return QStringLiteral( "vectorgeometry" );
```
```suggestion
  return QStringLiteral( "vectorcreation" );
```

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstolines.cpp:
```diff
@@ -0,0 +1,226 @@
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ),
+                QObject::tr( "Input layer" ), QList< int >() << QgsProcessing::TypeVectorPoint ) );
+  addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "CLOSE_LINES" ),
```
```suggestion
  addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "CLOSE_PATH" ),
```

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstolines.cpp:
```diff
@@ -0,0 +1,226 @@
+                QObject::tr( "Input layer" ), QList< int >() << QgsProcessing::TypeVectorPoint ) );
+  addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "CLOSE_LINES" ),
+                QObject::tr( "Create closed lines" ), false, true ) );
+  addParameter( new QgsProcessingParameterExpression( QStringLiteral( "ORDER_EXPRESSION" ),
```
Please add the older ORDER_FIELD compatibility code here from PointsToPaths.py also

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstolines.cpp:
```diff
@@ -0,0 +1,226 @@
+                QObject::tr( "Sort text containing numbers naturally" ), false, true ) );
+  addParameter( new QgsProcessingParameterField( QStringLiteral( "GROUP_FIELD" ),
+                QObject::tr( "Line group field" ), QVariant(), QStringLiteral( "INPUT" ), QgsProcessingParameterField::Any, false, true ) );
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ),
```
Also needs DATE_FORMAT parameter ported

uclaros replied:
@nyalldawson I am afraid this will need a complete `strptime` implementation, just for compatibility sake of a poorly documented parameter. Got any suggestion?

nyalldawson replied:
We could potentially make a converter from strptime format strings to the qt QDateTime::fromString formats, (see https://doc.qt.io/qt-5/qtime.html#fromString-1), at least for the common format parameters.

uclaros replied:
We could use `strptime` for linux builds and a map from strptime to qdatetime strings for windows, but that will not be very pretty and probably not fully backwards compatible! I'd really rather maintain the python code, but it's your call!

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstolines.cpp:
```diff
@@ -0,0 +1,226 @@
+  addParameter( new QgsProcessingParameterField( QStringLiteral( "GROUP_FIELD" ),
+                QObject::tr( "Line group field" ), QVariant(), QStringLiteral( "INPUT" ), QgsProcessingParameterField::Any, false, true ) );
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ),
+                QObject::tr( "Lines" ), QgsProcessing::TypeVectorLine ) );
```
Will also need the OUTPUT_TEXT_DIR handling added here too (although that should really have been a separate algorithm all along -- it's too late to change unfortunately)

uclaros commented:
@nyalldawson This is not a direct port of *Points to path*, it is supposed to be a new, replacement algorithm, as commented in https://github.com/qgis/QGIS/pull/39596#issuecomment-716221208.
The reasoning for doing so:
- Points to path is becoming complicated to be kept backwards compatible, forcing things like *order_expression* to be an optional but required parameter leading to user confusion (#39596).
- *order_field* logic is redundant since the *order_expression* is introduced.
- *date_format* logic is redundant since the *order_expression* is introduced.
- the outputting text part covers one very specific case and should not be in a generic *points -> linestrings* algorithm
- such an algorithm belongs to the *vector geometry* group together with *Polygonize* and *Polygons to lines* algorithms so it can be easily found.
- *Points to paths* is an unfortunate name choice for a generic *points -> linestrings* algorithm. People will keep asking *how can I convert points to lines?*.

Since #39415 has a 3.18 milestone, my suggestion is to revert it and keep *Points to path* as is, as a purpose built tool, marked as deprecated (is it possible?), kept for backwards compatibility with existing models. Then introduce this as a generic *points -> linestrings* algorithm which has both #39415 and #38701 functionality baked in.

uclaros committed:
Implement GROUP_EXPRESSION instead of GROUP_FIELD
to allow for flexible grouping without the need for an intermediate field calculation.

uclaros committed:
Update help string

nyalldawson commented:
> my suggestion is to revert it and keep Points to path as is, as a purpose built tool, marked as deprecated (is it possible?)

I'm honestly not a fan of that approach. There's no technical reason the port can't be made to be a drop in replacement, and leaving the python version just gives us a bunch of annoying (fragile) python code to drag around for the next couple of years :sob:

uclaros commented:
Well, if the python code is an annoyance then I could also port the original alg as it is in last release. IMHO it is a good opportunity to address the bullets in my previous comment, otherwise we'll have a bunch of annoying (redundant) c++ code to drag around for the next couple of years to come ;)

uclaros committed:
Const iterate to avoid copy; fix empty group expression case

alexbruy commented:
I'm with @nyalldawson here. There is no technical reasons why it can't be ported without breaking backward compatibility and intorducing yet another duplicated tool. We already did this for lot of algorithms. It is much better to have a single alg and not to confuse users.

uclaros commented:
I'm sorry, I believe the first three bullets are techical enough reasons to do so. Is there even a way for #39596 to be solved with a backwards compatible alg?
I think a new algorithm is the best solution here and so did @nyalldawson when I first suggested it in https://github.com/qgis/QGIS/pull/39596#issuecomment-716221208

alexbruy commented:
In that discussion there is an [explanation](https://github.com/qgis/QGIS/pull/39596#issuecomment-716358902) how it works keeping backward compatibility for existing scripts and models.

alexbruy commented:
Also it is not clear to me why "order expression" parameter has to be optional and why that idea came up at all. The "order field" was always required in this algorithm, so does the "order expression". The reason to make "order field" optional was its deprecation in favor of "order expression", this allows old scripts and models to set "order field" and continue to work, while new scripts and models have to use "order expression".

uclaros commented:
I'm sorry, it seems I didn't explain myself correctly. The problem I am referring to with this solution is that we have an **optional** parameter that **cannot be left blank**. How does that lead to less user confusion than having a second algorithm?

uclaros commented:
> Also it is not clear to me why "order expression" parameter has to be optional

"order_expression" needs to be optional for backwards compatibility

nyalldawson commented:
I think we could work around that with a new parameter definition flag "OptionalForBackwardCompatilibityOnly", which tells the gui to treat it the same as any other non optional parameters. (But allows for scripts and models to proceed even if it's not specified)

alexbruy commented:
Looking at other algs with deprecated parameters I see that everywhere we made deprecated parameter optional and replacement parameter mandatory with some sensible default. Then if deprecated parameter is set by an old script/model it is used instead of the new parameter. This way everything continues to work and this scenario tested by our test suite.

nyalldawson commented:
Manually marking as stale because stalebot has left for christmas break

uclaros commented:
Can we have #38701 merged until this is sorted out?

uclaros committed:
Change points to lines to points to paths

uclaros committed:
Complete points to paths port

uclaros committed:
Forgot begin-end fields

uclaros committed:
Fix parameter name

uclaros committed:
Reorder fields

uclaros commented:
OK, I modified this to completely replace the python version and be pretty much backwards compatible.
Possible incompatibilities:
1. `strptime` to `QDateTime` formats conversion is not 100% handled. Some uncommon date formats will not sort correctly in existing models.
2. The order of groups in the output file was undefined (dict iteration) and is still undefined but possibly different (QHash iteration)

I am not sure why the test fails. I see in https://github.com/qgis/QGIS/blob/8a3eefd0a599dd75b3a260c23a7eb5ccfae742c5/python/plugins/processing/tests/testdata/qgis_algorithm_tests2.yaml#L1588-L1601 there are two outputs. Is it supposed to be an OR? Couldn't find it in https://github.com/qgis/QGIS/blob/master/python/plugins/processing/tests/README.md

uclaros commented:
Any chance natural sorting for points to lines will not hit a second feature freeze?

nyalldawson commented:
it's on my list to try and review prior to freeze!

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+{
+  return QObject::tr( "This algorithm takes a point layer and connects its features creating a new line layer.\n\n"
+                      "An attribute or expression may be specified to define the order the points should be connected. "
+                      "If no order expression is specified, the fid is used.\n\n"
```
```suggestion
                      "If no order expression is specified, the feature ID is used.\n\n"
```

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+
+QString QgsPointsToPathsAlgorithm::group() const
+{
+  return QObject::tr( "Vector geometry" );
```
```suggestion
  return QObject::tr( "Vector creation" );
```

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+
+QString QgsPointsToPathsAlgorithm::groupId() const
+{
+  return QStringLiteral( "vectorgeometry" );
```
```suggestion
  return QStringLiteral( "vectorcreation" );
```

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+                QObject::tr( "Order expression" ), QVariant(), QStringLiteral( "INPUT" ), true ) );
+  addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "NATURAL_SORT" ),
+                QObject::tr( "Sort text containing numbers naturally" ), false, true ) );
+  addParameter( new QgsProcessingParameterExpression( QStringLiteral( "GROUP_EXPRESSION" ),
```
The existing algorithm has "GROUP_FIELD" parameter which also needs to be retained for compatibility

uclaros replied:
Seems like I subconsciously skipped this, since I didn't like the idea of the complete port!! :laughing:

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+  if ( ! groupExpressionString.isEmpty() )
+  {
+    requiredFields.append( groupExpression.referencedColumns().values() );
+    const QgsField field = groupExpression.isField() ? source->fields().field( requiredFields.last() ) : QString( "group" );
```
```suggestion
    const QgsField field = groupExpression.isField() ? source->fields().field( requiredFields.last() ) : QStringLiteral( "group" );
```

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+    throw QgsProcessingException( QObject::tr( "The text output directory does not exist" ) );
+
+  // Store the points in a hash with the group identifier as the key
+  QHash< QVariant, QVector< QPair< QVariant, const QgsPoint * > > > allPoints;
```
```suggestion
  QHash< QVariant, QVector< QPair< QVariant, QgsPoint > > > allPoints;
```

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+    {
+      break;
+    }
+    feedback->setProgress( currentPoint * totalPoints );
```
```suggestion
    feedback->setProgress( 0.5 * currentPoint * totalPoints );
```

so that this first iteration only reports progress up to 50%

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+    }
+    feedback->setProgress( currentPoint * totalPoints );
+
+    if ( f.hasGeometry() && ! f.geometry().isNull() )
```
```suggestion
    if ( f.hasGeometry() )
```

the null check is redundant -- it's what hasGeometry does

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+      const QVariant orderValue = orderExpression.evaluate( &expressionContext );
+      const QVariant groupValue = groupExpressionString.isEmpty() ? QVariant() : groupExpression.evaluate( &expressionContext );
+
+      if ( ! allPoints.keys().contains( groupValue ) )
```
```suggestion
      if ( ! allPoints.contains( groupValue ) )
```

(otherwise all keys are converted to a list, which will be slow for large hashes, and then that list will be searched for the value (also very slow))

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+      const QVariant groupValue = groupExpressionString.isEmpty() ? QVariant() : groupExpression.evaluate( &expressionContext );
+
+      if ( ! allPoints.keys().contains( groupValue ) )
+        allPoints[ groupValue ] = QVector< QPair< QVariant, const QgsPoint * > >();
```
```suggestion
        allPoints[ groupValue ] = QVector< QPair< QVariant, QgsPoint > >();
```

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+
+      if ( ! allPoints.keys().contains( groupValue ) )
+        allPoints[ groupValue ] = QVector< QPair< QVariant, const QgsPoint * > >();
+      const QgsPoint *point = qgsgeometry_cast< const QgsPoint * >( f.geometry().constGet()->clone() );
```
```suggestion
      QgsPoint point( *qgsgeometry_cast< const QgsPoint * >( f.geometry().constGet() ) ) );
```

But warning -- this will crash if a multipoint source is used, so we should handle the case here that the feature's geometry is multipoint and add all those points if so

uclaros replied:
ok, I also switched to `std::stable_sort` so the point order within a multipoint is preserved; we have no way to manipulate it since the points share attributes.

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+      if ( ! allPoints.keys().contains( groupValue ) )
+        allPoints[ groupValue ] = QVector< QPair< QVariant, const QgsPoint * > >();
+      const QgsPoint *point = qgsgeometry_cast< const QgsPoint * >( f.geometry().constGet()->clone() );
+      allPoints[ groupValue ] << QPair< QVariant, const QgsPoint *>( orderValue, point );
```
```suggestion
      allPoints[ groupValue ] << qMakePair( orderValue, point );
```

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+
+  int pathCount = 0;
+  currentPoint = 0;
+  QHashIterator< QVariant, QVector< QPair< QVariant, const QgsPoint * > > > hit( allPoints );
```
```suggestion
  QHashIterator< QVariant, QVector< QPair< QVariant, QgsPoint > > > hit( allPoints );
```

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+    {
+      std::sort( pairs.begin(),
+                 pairs.end(),
+                 [&collator]( const QPair< const QVariant, const QgsPoint * > &pair1,
+                              const QPair< const QVariant, const QgsPoint * > &pair2 )
```
```suggestion
                 [&collator]( const QPair< const QVariant, QgsPoint > &pair1,
                              const QPair< const QVariant, QgsPoint > &pair2 )
```

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+    {
+      std::sort( pairs.begin(),
+                 pairs.end(),
+                 []( const QPair< const QVariant, const QgsPoint * > &pair1,
+                     const QPair< const QVariant, const QgsPoint * > &pair2 )
```
```suggestion
                 []( const QPair< const QVariant, QgsPoint > &pair1,
                     const QPair< const QVariant, QgsPoint > &pair2 )
```

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+
+
+    QVector<QgsPoint> pathPoints;
+    for ( QVector< QPair< QVariant, const QgsPoint * > >::ConstIterator pit = pairs.constBegin(); pit != pairs.constEnd(); ++pit )
```
```suggestion
    for ( auto pit = pairs.constBegin(); pit != pairs.constEnd(); ++pit )
```

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+      {
+        break;
+      }
+      feedback->setProgress( currentPoint * totalPoints );
```
This feedback report should be done up a level, in the loop over the groups (ie. report progress as each output feature is written)

uclaros replied:
I did that on purpose, so that we have a smoother progress bar regardless of the number of groups. If I put it up a level, then a million groups of a few points each will look smooth, but a few groups with a million points each will not!

nyalldawson replied:
Hm -- i guess ideally we'd do it on both levels -- so we'd report the feedback on the group, and then on the points within the group. If you want to try this QgsProcessingMultiStepFeedback can be used, which allows you to specify the total number of steps and then the individual progress reports are scaled accordingly.

uclaros replied:
If groups have uneven number of points then the progress bar will not have constant speed (think windows saying your copy will take 3 minutes but stays at 99% for 2 hours!) I think that reporting on a per point basis guarantees that the progress bar shows the correct fraction of the total time the process will take. Again, this only matters when points are by the millions, otherwise the progress bar is shown for too little to matter!

nyalldawson replied:
Ah, I read the code wrong -- you're approach is good! It just needs scaling so it only updates the final 50% of the progress instead of restarting at 0

uclaros replied:
I did that when I changed the progress bar on the first iteration as per your suggestion!

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+        break;
+      }
+      feedback->setProgress( currentPoint * totalPoints );
+      pathPoints.append( *pit->second );
```
```suggestion
      pathPoints.append( pit->second );
```

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+      feedback->pushInfo( QObject::tr( "Skipping path with group %1 : insufficient vertices" ).arg( hit.key().toString() ) );
+      continue;
+    }
+    if ( closePaths && pathPoints.size() > 2 && pathPoints.first() != pathPoints.last() )
```
```suggestion
    if ( closePaths && pathPoints.size() > 2 && pathPoints.constFirst() != pathPoints.constLast() )
```

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+      continue;
+    }
+    if ( closePaths && pathPoints.size() > 2 && pathPoints.first() != pathPoints.last() )
+      pathPoints.append( pathPoints.first() );
```
```suggestion
      pathPoints.append( pathPoints.constFirst() );
```

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+    QgsFeature outputFeature;
+    QgsAttributes attrs;
+    if ( ! groupExpressionString.isEmpty() )
+      attrs.append( hit.key() );
```
else append a QVariant(), or the attributes will go out-of-sync with the fields

uclaros replied:
Fields are handled similarly (group field is skipped if unneeded, this is how the python version works):
https://github.com/qgis/QGIS/blob/1afd4048e985c81320b3ce3c15bdb64920fcc523/src/analysis/processing/qgsalgorithmpointstopaths.cpp#L183-L191

nyalldawson replied:
Ah, missed that -- thanks!

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,334 @@
+      if ( !textFile.open( QIODevice::WriteOnly | QIODevice::Text ) )
+        throw QgsProcessingException( QObject::tr( "Cannot open file for writing " ) + filename );
+
+      QgsDistanceArea da = QgsDistanceArea();
+      da.setSourceCrs( source->sourceCrs(), context.transformContext() );
+      da.setEllipsoid( context.ellipsoid() );
```
This construction is expensive, so should happen outside of the feature loop

nyalldawson reviewed (commented):
Looking good! A few minor fixes, but the big requirement is handling the old GROUP_FIELD parameter like we've done with ORDER_FIELD

nyalldawson commented:
Marking as freeze exempt for a short period given the original submission date and lengthy review delays

uclaros committed:
Address review comments

uclaros committed:
Handle multipoint input

uclaros committed:
Handle compatibility parameter GROUP_FIELD

uclaros committed:
Fix failing test
Output feature order cannot be made backwards compatible so we're forced to modify the test to pass.

nyalldawson commented on src/analysis/processing/qgsalgorithmpointstopaths.cpp:
```diff
@@ -0,0 +1,358 @@
+      QObject::tr( "Order field" ), QVariant(), QString(), QgsProcessingParameterField::Any, false, true );
+  orderField->setFlags( orderField->flags() | QgsProcessingParameterDefinition::FlagHidden );
+  addParameter( orderField );
+  QgsProcessingParameterField *groupField = new QgsProcessingParameterField( QStringLiteral( "GROUP_FIELD" ),
+      QObject::tr( "Group field" ), QVariant(), QString(), QgsProcessingParameterField::Any, false, true );
```
```suggestion
  QgsProcessingParameterField *groupField = new QgsProcessingParameterField( QStringLiteral( "GROUP_FIELD" ),
      QObject::tr( "Group field" ), QVariant(), QStringLiteral( "INPUT" ), QgsProcessingParameterField::Any, false, true );
```

uclaros replied:
Same goes for deprecated `ORDER_FIELD` parameter. But I thought it doesn't matter since it's hidden and used only by scripts.

nyalldawson committed:
Update src/analysis/processing/qgsalgorithmpointstopaths.cpp

roya0045 commented:
@uclaros 👍

uclaros commented:
Thank you for the freeze exempt!

nyalldawson commented:
@uclaros looks like the tests are randomly failing for this one -- see eg https://travis-ci.org/github/qgis/QGIS/builds/755098123

uclaros commented:
@nyalldawson This is most probably a matter of feature order on the output layer, because we cannot guarantee the output group order since it comes from iterating the hash. Is there a proper way to tell the test that feature order does not matter? If not, I can add some sorting step on the test. Another possible solution would be to explicitly sort the output layer by the group field-expression so we can guarantee the output order, but that could unnecessarily slow things down on huge datasets.

agiudiceandrea commented:
@uclaros, please see issue https://github.com/qgis/QGIS/issues/54686.
